src/responder.ts

```diff
@@ -51,182 +61,357 @@ enum ResponderEvent {
-    private timeLastBlockReceived: number = Date.now();
-    private mAttemptsDone: number = 0;
+    private timeLastBlockReceived: number;
 
-    get attemptsDone() {
-        return this.mAttemptsDone;
-    }
+    private transactionMiner: EthereumTransactionMiner;
 
     /**
      * @param signer The signer of the wallet associated with this responder. Each responder should have exclusive access to his wallet.
-     * @param appointmentId The id of the Appointment this object is responding to.
-     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
-     * @param [confirmationsRequired=40] The number of confirmations required before a transaction is trusted.
-     * @param [maxAttempts=10] The maximum number of retries before the Responder will give up.
+     * @param [confirmationsRequired] The number of confirmations required before a transaction is trusted.
+     * @param [maxAttempts] The maximum number of retries before the Responder will give up.
      */
     constructor(
-        readonly signer: ethers.Signer,
-        public readonly appointmentId: string,
-        public readonly ethereumResponse: IEthereumResponse,
-
-        public readonly confirmationsRequired = 40,
-        private readonly maxAttempts: number = 10
+        signer: ethers.Signer,
+        private readonly gasPolicy: GasPolicy,
+        public readonly confirmationsRequired: number,
+        private readonly maxAttempts: number,
+        transactionMiner?: EthereumTransactionMiner
     ) {
-        super(signer, appointmentId, ethereumResponse);
-
-        signer.provider.on("block", this.updateLastBlockTime);
+        super(signer);
+
+        this.transactionMiner = transactionMiner || new EthereumTransactionMiner(
+            this.signer,
+            this.confirmationsRequired,
+            EthereumDedicatedResponder.WAIT_BLOCKS_BEFORE_RETRYING,
+            EthereumDedicatedResponder.WAIT_TIME_FOR_NEW_BLOCK
+        );
     }
 
-    /**
-     * Release any resources used by this instance.
-     */
-    private destroy() {
-        this.signer.provider.removeListener("block", this.updateLastBlockTime);
+    // Makes sure that the class is locked while `fn` is running, and that any listener is registered and cleared correctly
+    private async withLock(fn: () => Promise<any>) {
+        if (this.locked) {
+            throw new ApplicationError("This responder can ony handle one response at a time."); // TODO:93: more specific Error type?
+        }
+
+        this.locked = true;
+
+        const listener = this.newBlockReceived.bind(this);
+        this.signer.provider.on("block", listener);
+
+        this.timeLastBlockReceived = Date.now();
+
+        try {
+            await fn();
+        } finally {
+            this.signer.provider.removeListener("block", listener);
+            this.locked = false;
+        }
     }
 
-    private updateLastBlockTime() {
+    private newBlockReceived(blockNumber: number) {
```

Comment:
The blockNumber param here isnt used

Reply:
Removed in 352fc270d00d7838ef20e737c027295a3bb1c97d.